source/kernel/bank2/fat.mac

```diff
@@ -915,6 +945,45 @@ DTY_ADD:
 	ret	nz
 	dec	hl
 	ret
+;
+;
+;------------------------------------------------------------------------------
+;
+	PROC	ACLU_SET
+;
+;    Sets the UD_ACLU field of a unit descriptor.
+;
+; Entry:   DE = Value to set
+;	       HL = Address of unit descriptor
+; Returns:     None
+; Corrupts:    None
+;
+	push ix	
+	push hl
+	pop ix
+	ld (ix+UD_ACLU##),e
+	ld (ix+UD_ACLU##+1),d
+	pop ix
+	ret
+;
+;
+;------------------------------------------------------------------------------
+;
+;    Gets the UD_ACLU field of a unit descriptor.
+;
+; Entry:   HL = Address of unit descriptor
+; Returns: DE = Value
+; Corrupts:    None
+;
+ACLU_GET:
+	push ix	
+	push hl
+	pop ix
+	ld e,(ix+UD_ACLU##)
+	ld d,(ix+UD_ACLU##+1)
+	pop ix

```

Comment:
I would go with

    push hl 
    push bc 
    ld bc, UD_ACLU## 
    add hl, bc  
    ld e, (hl)  
    inc hl  
    ld d, (hl)  
    pop bc 
    pop hl 

I had calculated bytes and cycles (including that famous one extra wait for every command) and if i'm not mistaken, avoiding use of IX would make code 2 bytes shorted and 17 cycles faster. I guess ACLU_SET would benefit from similar change too

Reply:
Well, missed that my version will corrupt AF unfortunately. Sorry for noise

Reply:
What if we `push af` and `pop af` then? The code wouldn't be shorter but we would still gain some cycles, right?

Reply:
What about using DE instead of BC in @Pencioner version?:

```
	push	af
	push	hl
	ld	de, UD_ACLU##
	add	hl, de
	ld	e, (hl)
	inc	hl
	ld	d, (hl)
	pop	hl
	pop	af
; 92 t-states (-11), 11 bytes (-2)
```

If I'm not mistaken:
- Original code: 103 t-states, 13 bytes
- Pencioner proposal: 92 t-states (-11), 11 bytes (-2) _(but corrupts F, as stated)_
- Pencioner proposal with `push af` and `pop af`: 115 t-states (+12), 13 bytes
- Pencioner proposal using DE: 92 t-states (-11), 11 bytes (-2)

I'm not familiar with Nextor code, so I don't know if it is required not to corrupt F... If the `push af` and `pop af` can be removed, the routine is: 69 t-states (-34), 9 bytes (-4)

Reply:
`ACLU_GET` is called in two places, one needs AF to be preserved and the other doesn't, so I guess the responsibility of preserving AF could be moved to the caller. And for sure a similar optimization could be done for `ACLU_SET` too; I'll take a look, thank you very much!

Reply:
Moving the responsability of preserving AF to the callers sounds good.

Are these those two places?
- https://github.com/Konamiman/Nextor/blob/improved-free-cluster-search/source/kernel/bank2/fat.mac#L656
- https://github.com/Konamiman/Nextor/blob/improved-free-cluster-search/source/kernel/bank2/fat.mac#L891

I'm trying to find reads of A or the flags before being overwritten after calling ACLU_GET and I can't find them...

Reply:
Yep, these two. And for the case of `AL_CLUSTERS`, a parameter is passed in A but indeed it's stored for later usage before `ACLU_GET` is called, so looks like it's not actually necessary to preserve it or the flags.